src/main/java/org/wso2/carbon/hash/provider/pbkdf2/PBKDF2HashProvider.java

```diff
@@ -0,0 +1,166 @@
+                            ErrorMessage.ERROR_CODE_EMPTY_VALUE.getCode());
+        }
+    }
+
+    /**
+     * This method is responsible for validating and getting iteration count from the metaProperties.
+     *
+     * @param metaProperties The Map which has the properties which needs for PBKDF2 hashing algorithm.
+     * @return the iteration count which needs to be get.
+     */
+    private int getCount(Map<String, Object> metaProperties) {
+
+        if (metaProperties.get(Constants.ITERATION_COUNT_KEY) == null) {
+            return Constants.DEFAULT_ITERATION_COUNT;
+        }
+        return (int) metaProperties.get(Constants.ITERATION_COUNT_KEY);
+    }
+
+    /**
+     * This method is responsible for validating and getting derived key length from the metaProperties.
+     *
+     * @param metaProperties The Map which has the properties which needs for PBKDF2 hashing algorithm.
+     * @return the derived key length which needs to be get.
+     */
+    private int getDerivedKeyLength(Map<String, Object> metaProperties) {
+
+        if (metaProperties.get(Constants.DERIVED_KEY_LENGTH_KEY) == null) {
+            return Constants.DEFAULT_DERIVED_KEY_LENGTH;
+        }
+        return (int) metaProperties.get(Constants.DERIVED_KEY_LENGTH_KEY);
+    }
+
+    /**
+     * This method is responsible for validating and getting Pseudo Random Function from the metaProperties.
+     *
+     * @param metaProperties The Map which has the properties which needs for PBKDF2 hashing algorithm.
+     * @return the pseudo random function which needs to be get.
+     */
+    private String getPseudoRandomFunction(Map<String, Object> metaProperties) {
+
+        if (metaProperties.get(Constants.PSEUDO_RANDOM_FUNCTION_KEY) == null) {
+            return Constants.DEFAULT_PBKDF2_PRF;
+        }
+        return (String) metaProperties.get(Constants.PSEUDO_RANDOM_FUNCTION_KEY);
+    }
+
+    /**
+     * This method is responsible for the implementation of PBKDF2 hashing algorithm.
+     *
+     * @param value                The value (eg:- Password, token) which needs to be hashed.
+     * @param salt                 The salt value for each respective values.
+     * @param iterationCount       Iteration count denotes how iteratively the value needs to be hashed inside PRF.
+     * @param dkLen                The output length of the hash function.
+     * @param pseudoRandomFunction the PRF which needs to be used in PBKDF2 hashing.
+     * @return The resulting hash value of the value.
+     * @throws HashProviderException This exception includes the exceptional handling for pbkdf2 hashing.
+     */
+    private byte[] pbkdf2HashCalculation(String value, String salt, int iterationCount, int dkLen,
+                                         String pseudoRandomFunction)
+            throws HashProviderException {
+
+        try {
+            PBEKeySpec spec = new PBEKeySpec(value.toCharArray(), base64ToByteArray(salt), iterationCount, dkLen);
```

Comment:
How are we handling a salt being null?

Reply:
```
 if (salt == null) {
            throw new NullPointerException("the salt parameter must be non-null");
        } else if (salt.length == 0) {
            throw new IllegalArgumentException("the salt parameter must not be empty");
        }
```
handling inside the PBEKeySpec class.

Reply:
We should not throw any null pointers or any other exceptions from this service. Shall we catch them and throw specific exception types?